gson/src/main/java/com/google/gson/internal/bind/TypeAdapterRuntimeTypeWrapper.java

```diff
@@ -51,7 +52,7 @@ public void write(JsonWriter out, T value) throws IOException {
 
     TypeAdapter chosen = delegate;
     Type runtimeType = getRuntimeTypeIfMoreSpecific(type, value);
-    if (runtimeType != type) {
+    if (runtimeType != type && !Primitives.areEquivalentTypes(runtimeType, type)) {
       TypeAdapter runtimeTypeAdapter = context.getAdapter(TypeToken.get(runtimeType));

```

Comment:
What I intended to say was that this line will do the primitive/wrapper work automatically so the check above is unnecessary.